modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceCallContextImpl.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.service;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.ignite.internal.util.typedef.internal.A;
+import org.apache.ignite.services.ServiceCallContext;
+
+/**
+ * Service call context implementation.
+ */
+public class ServiceCallContextImpl implements ServiceCallContext {
+    /** Service call context attributes. */
+    private final Map<String, byte[]> attrs;
+
+    /**
+     * Default contructor.
+     */
+    public ServiceCallContextImpl() {
+        attrs = new HashMap<>();
+    }
+
+    /**
+     * Constructs an immutable context from the map.
+     *
+     * @param attrs Service call context attributes.
+     */
+    public ServiceCallContextImpl(Map<String, byte[]> attrs) {
+        this.attrs = Collections.unmodifiableMap(attrs);
+    }
+
+    /**
+     * Constructs an immutable context.
+     *
+     * @param callCtx Service call context.
+     */
+    public ServiceCallContextImpl(ServiceCallContext callCtx) {
```

Comment:
I suppose that it is a really bad idea.

Suppose that someone pass another implementation of ServiceCallContext?

Is it ok to fail in runtime in this case?